edgelet/iotedge-proxy/src/lib.rs

```diff
@@ -0,0 +1,163 @@
+    use openssl::pkey::PKey;
+    use openssl::rsa::Rsa;
+    use openssl::x509::extension::{
+        AuthorityKeyIdentifier, BasicConstraints, ExtendedKeyUsage, KeyUsage, SubjectKeyIdentifier,
+    };
+    use openssl::x509::{X509Name, X509};
+    use std::fmt::{Display, Formatter};
+    use std::io::Error as IoError;
+
+    #[derive(Debug, Fail)]
+    pub enum CertGeneratorError {
+        ErrorStack(ErrorStack),
+        Io(IoError),
+    }
+
+    impl Display for CertGeneratorError {
+        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+            match self {
+                CertGeneratorError::ErrorStack(err) => write!(f, "{}", err),
+                CertGeneratorError::Io(err) => write!(f, "{}", err),
+            }
+        }
+    }
+
+    impl From<ErrorStack> for CertGeneratorError {
+        fn from(err: ErrorStack) -> Self {
+            CertGeneratorError::ErrorStack(err)
+        }
+    }
+
+    impl From<IoError> for CertGeneratorError {
+        fn from(err: IoError) -> Self {
+            CertGeneratorError::Io(err)
+        }
+    }
+
+    #[derive(Default)]
+    pub struct CertGenerator {
+        public_key: Option<PathBuf>,
+        private_key: Option<PathBuf>,
+        cert: Option<PathBuf>,
+        common_name: Option<String>,
+    }
+
+    impl CertGenerator {
+        pub fn private_key(&mut self, path: &Path) -> &Self {
+            self.private_key = Some(path.to_path_buf());
+            self
+        }
+
+        pub fn public_key(&mut self, path: &Path) -> &Self {
+            self.public_key = Some(path.to_path_buf());
+            self
+        }
+
+        pub fn cert(&mut self, path: &Path) -> &Self {
+            self.cert = Some(path.to_path_buf());
+            self
+        }
+
+        pub fn common_name(&mut self, name: String) -> &Self {
+            self.common_name = Some(name);
+            self
+        }
+
+        pub fn generate(&self) -> Result<X509, CertGeneratorError> {
+            let rsa = Rsa::generate(2048)?;
+            let pkey = PKey::from_rsa(rsa)?;
+
+            if let Some(pkey_path) = &self.public_key {
+                fs::write(pkey_path, pkey.public_key_to_pem()?)?;
+            }
+
+            let mut name = X509Name::builder()?;
+            name.append_entry_by_nid(
+                Nid::COMMONNAME,
+                self.common_name
+                    .as_ref()
+                    .unwrap_or(&"localhost".to_string()),
+            )?;
+            let name = name.build();
+
+            let mut builder = X509::builder()?;
+            builder.set_version(2)?;
+            builder.set_subject_name(&name)?;
+            builder.set_issuer_name(&name)?;
+            builder.set_not_before(Asn1Time::days_from_now(0)?.as_ref())?;
+            builder.set_not_after(Asn1Time::days_from_now(365)?.as_ref())?;
+            builder.set_pubkey(&pkey)?;
+
+            let basic_constraints = BasicConstraints::new().critical().ca().build()?;
```

Comment:
Can you comment some of this? Like it would be good to know what is happening here. Example - what is "critical"?

Reply:
I have plans to add integration test and I think some parts of this code will change.

Reply:
I still think that some comments might be useful :) Even if just talking about major chunks. What if our DRI needs to debug something off hours?